### PR TITLE
Use hashbrown replacements for std equivalents

### PR DESCRIPTION
**Pull request by GeneFerneau**

### CODE REVIEW

File: lightning/Cargo.toml [hunk 1]
```diff
@@ -29,6 +29,7 @@ unstable = []
 [dependencies]
 bitcoin = "0.26"
 
+hashbrown = { version = "0.9" }
```
Comment (GeneFerneau) [inline lightning/Cargo.toml:32]: Current `hashbrown` is `0.11.2`, can bump to `0.11`, if MSRV `1.49` is OK

Can also feature-gate inside the prelude module, too.

**Comment by TheBlueMatt** [comment]
Can we make this optional instead with a "hashbrown" feature (even an implicit one where you just make "hashbrown" an optional dependency)? Then we can make no-std depend on it.

File: lightning/src/routing/router.rs [hunk 2]
```diff
@@ -3838,16 +3837,19 @@ mod tests {
 		}
 	}
 
+    #[cfg(feature = "std")]
 	pub(super) fn random_init_seed() -> u64 {
 		// Because the default HashMap in std pulls OS randomness, we can use it as a (bad) RNG.
-		use core::hash::{BuildHasher, Hasher};
```
Comment (devrandom) [inline lightning/src/routing/router.rs:3843]: indentation here and below is off

File: lightning/src/lib.rs [hunk 3]
```diff
@@ -44,5 +44,12 @@ pub mod ln;
 pub mod routing;
 
 mod prelude {
-	pub use alloc::{vec, vec::Vec, string::String};
-}
\ No newline at end of file
+    #[cfg(feature = "no_std")]
+    extern crate hashbrown;
+
+	pub use alloc::{vec, vec::Vec, string::String, collections::VecDeque};
```
Comment (devrandom) [inline lightning/src/lib.rs:50]: indentation is off

**Review by devrandom — APPROVED** [review]
ACK other than nits

File: lightning/Cargo.toml [hunk 4]
```diff
@@ -25,10 +26,13 @@ max_level_debug = []
 # This is unsafe to use in production because it may result in the counterparty publishing taking our funds.
 unsafe_revoked_tx_signing = []
 unstable = []
+std = []
```
Comment (TheBlueMatt) [inline lightning/Cargo.toml:29]: Is there a reason to have an std feature at all?

File: lightning/src/lib.rs [hunk 5]
```diff
@@ -44,5 +44,12 @@ pub mod ln;
 pub mod routing;
 
 mod prelude {
-	pub use alloc::{vec, vec::Vec, string::String};
-}
\ No newline at end of file
+    #[cfg(feature = "no_std")]
```
Comment (TheBlueMatt) [inline lightning/src/lib.rs:47]: I think we should directly check for the `hashbrown` feature. Ideally we'd be able to use `hashbrown` with `std` or `no_std`, but require it for `no_std`.

Comment (GeneFerneau) [reply lightning/src/lib.rs:47]: Hrm, I tried directly naming a `hashbrown` feature, but I think it collides with the dependency name. Maybe I'm doing something wrong?

Comment (TheBlueMatt) [reply lightning/src/lib.rs:47]: Huh, I thought if you don't even write out the feature and just make the dependency optional you can use it as a feature in cfg.

Comment (GeneFerneau) [reply lightning/src/lib.rs:47]: Ok, maybe that's what I was doing wrong. I tried writing out `hashbrown = ["hashbrown"]` as a feature. Will try what you described.

File: ci/check-compiles.sh [hunk 6]
```diff
@@ -6,3 +6,4 @@ cargo check
 cargo doc
 cargo doc --document-private-items
 cd fuzz && cargo check --features=stdin_fuzz
+cd ../lightning && cargo check --no-default-features --features=no_std
```
Comment (TheBlueMatt) [inline ci/check-compiles.sh:9]: In `.github/workflows/build.yml` can you also add a `cd lightning && cargo test --verbose --color always --features hashbrown` to run the test suite with hashbrown + std?

File: lightning/src/routing/router.rs [hunk 7]
```diff
@@ -3838,16 +3837,19 @@ mod tests {
...
 
+	#[cfg(not(feature = "no_std"))]
 	pub(super) fn random_init_seed() -> u64 {
 		// Because the default HashMap in std pulls OS randomness, we can use it as a (bad) RNG.
 		use core::hash::{BuildHasher, Hasher};
 		let seed = std::collections::hash_map::RandomState::new().build_hasher().finish();
 		println!("Using seed of {}", seed);
 		seed
 	}
+    #[cfg(not(feature = "no_std"))]
```
Comment (TheBlueMatt) [inline lightning/src/routing/router.rs:3848]: nit: indentation with tabs.

**Review by TheBlueMatt — APPROVED** [review]

File: ci/check-compiles.sh (hunk 6 above)
Comment (GeneFerneau) [reply ci/check-compiles.sh:9]: Which builds should that run under, or do you want a new build target?

File: lightning/src/routing/router.rs (hunk 7 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:3848]: If there's an easy way to just swap spaces for tabs, sure (I don't think there is, though?), but we've tried rustfmt a few times and it makes tons of code completely unreadable. It's desire to lay out all your code completely vertically means you lose a ton of context when reviewing more often than not. There's a few issues open upstream to make it more sensible, but I don't believe there's been much progress on any of them.

File: ci/check-compiles.sh (hunk 6 above)
Comment (TheBlueMatt) [reply ci/check-compiles.sh:9]: Uhhmmmmm, yes? I don't think it matters too much, maybe in the existing ones is simplest, given it's ~one line?

File: lightning/src/routing/router.rs (hunk 7 above)
Comment (GeneFerneau) [reply lightning/src/routing/router.rs:3848]: Ok, I won't worry about it then. Thought there might be a simple fix.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -56,6 +56,8 @@ jobs:
           cargo build --verbose --color always --features rpc-client
           cargo build --verbose --color always --features rpc-client,rest-client
           cargo build --verbose --color always --features rpc-client,rest-client,tokio
+          cd ../lightning
+          cargo build --verbose --color always --features hashbrown
```
Comment (TheBlueMatt) [inline .github/workflows/build.yml:60]: We should also test with `hashbrown` in the "! matrix.build-net-tokio" block immediately above this one.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -92,6 +98,8 @@ jobs:
           RUSTFLAGS="-C link-dead-code" cargo test --verbose --color always --features rpc-client
           RUSTFLAGS="-C link-dead-code" cargo test --verbose --color always --features rpc-client,rest-client
           RUSTFLAGS="-C link-dead-code" cargo test --verbose --color always --features rpc-client,rest-client,tokio
+          cd ../lightning
+          RUSTFLAGS="-C link-dead-code" cargo test --verbose --color always --features hashbrown
```
Comment (TheBlueMatt) [inline .github/workflows/build.yml:102]: I dunno if we want the last test we run in "matrix.coverage" (ie the binaries which we generate coverage reports from) to be hashbrown ones - maybe move this up above (or as the first command in) "Test on Rust ${{ matrix.toolchain }} with net-tokio and full code-linking for coverage generation"?

**Review by TheBlueMatt — APPROVED** [review]

File: .github/workflows/build.yml (hunk 9 above)
Comment (GeneFerneau) [reply .github/workflows/build.yml:102]: SGTM, wasn't sure, so just added it after the other tests.

Comment (GeneFerneau) [reply .github/workflows/build.yml]: hrm, seems like this causes build failure:

```
error[E0658]: use of unstable library feature 'ptr_cast'
   --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.11.2/src/raw/mod.rs:504:57
    |
504 |         NonNull::new_unchecked(self.table.ctrl.as_ptr().cast())
    |                                                         ^^^^
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/60602
```

Comment (GeneFerneau) [reply .github/workflows/build.yml]: Might only make sense to use `hashbrown` in later versions of Rust, per [rust-lang/rust#60602](https://github.com/rust-lang/rust/issues/60602#issuecomment-594005876)

Comment (TheBlueMatt) [reply .github/workflows/build.yml]: Right, it needs to only run on non-1.36 versions of rust. Maybe add this:
```
      - name: Test no_std builds Rust ${{ matrix.toolchain }}
        if: "matrix.build-net-tokio"
        run: cargo test --verbose --color always --features hashbrown
 ```

Comment (TheBlueMatt) [reply .github/workflows/build.yml]: > MSRV of hashbrown 0.11.2 is rust 1.49

Oh, I missed this comment, I guess we need the if there to just be something new that matches on stable instaed.

Comment (TheBlueMatt) [reply .github/workflows/build.yml]: Try:
```
diff --git a/.github/workflows/build.yml b/.github/workflows/build.yml
index 231db705..2f14a4c2 100644
--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -14,18 +14,27 @@ jobs:
                      # 1.41.0 is Debian stable
                      1.41.0,
                      # 1.45.2 is MSRV for lightning-net-tokio, lightning-block-sync, and coverage generation
-                     1.45.2]
+                     1.45.2,
+                     # 1.49 is MSRV for no_std
+                     1.49.0]
         include:
           - toolchain: stable
             build-net-tokio: true
+            build-no-std: true
           - toolchain: stable
             platform: macos-latest
             build-net-tokio: true
+            build-no-std: true
           - toolchain: stable
             platform: windows-latest
             build-net-tokio: true
+            build-no-std: true
           - toolchain: beta
             build-net-tokio: true
+            build-no-std: true
+          - toolchain: 1.45.2
+            build-net-tokio: true
+            build-no-std: true
           - toolchain: 1.45.2
             build-net-tokio: true
             coverage: true
@@ -76,6 +85,9 @@ jobs:
           cd ../lightning
           RUSTFLAGS="-C link-dead-code" cargo build --verbose --color always --features hashbrown
           cd ..
+      - name: Test no_std builds Rust ${{ matrix.toolchain }}
+        if: "matrix.build-no-std"
+        run: cargo test --verbose --color always --features hashbrown
       - name: Test on Rust ${{ matrix.toolchain }} with net-tokio
         if: "matrix.build-net-tokio && !matrix.coverage"
         run: cargo test --verbose --color always
```

Comment (GeneFerneau) [reply .github/workflows/build.yml]: I added a check `build-no-std`, and a `1.49.0` to the toolchain. It took a little experimentation, but I think I understand the CI build scripts better now.

The fuzz build seems to be failing for other reasons...

**Review by TheBlueMatt — APPROVED** [review]
Looks good, only diff since devrandom's ack at d9ca4ea are CI and comment addressing, fuzz test failure is due to a libsecp bump that exists on current HEAD anyway, so shouldn't hold this up.

**Pull request merged**